Samriddhi Verma
16BCE1375
Machine Learning Lab

# Importing Libraries

In [4]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

wine = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data', names = ["Cultivator", "Alchol", "Malic_Acid", "Ash", "Alcalinity_of_Ash", "Magnesium", "Total_phenols", "Falvanoids", "Nonflavanoid_phenols", "Proanthocyanins", "Color_intensity", "Hue", "OD280", "Proline"])

# Data Analysis

In [5]:
wine.head()

Cultivator  Alchol  Malic_Acid   Ash  Alcalinity_of_Ash  Magnesium  \
0           1   14.23        1.71  2.43               15.6        127   
1           1   13.20        1.78  2.14               11.2        100   
2           1   13.16        2.36  2.67               18.6        101   
3           1   14.37        1.95  2.50               16.8        113   
4           1   13.24        2.59  2.87               21.0        118   

   Total_phenols  Falvanoids  Nonflavanoid_phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color_intensity   Hue  OD280  Proline  
0             5.64  1.04   3.92     1065  
1             4.38  1.05   3.40     1050  
2             5.68  1.03   3.17     1185  
3             7.80  0.86   3.45     1480  
4             4.32  1.04   2.93      735

In [6]:
wine.describe().transpose()

count        mean         std     min       25%  \
Cultivator            178.0    1.938202    0.775035    1.00    1.0000   
Alchol                178.0   13.000618    0.811827   11.03   12.3625   
Malic_Acid            178.0    2.336348    1.117146    0.74    1.6025   
Ash                   178.0    2.366517    0.274344    1.36    2.2100   
Alcalinity_of_Ash     178.0   19.494944    3.339564   10.60   17.2000   
Magnesium             178.0   99.741573   14.282484   70.00   88.0000   
Total_phenols         178.0    2.295112    0.625851    0.98    1.7425   
Falvanoids            178.0    2.029270    0.998859    0.34    1.2050   
Nonflavanoid_phenols  178.0    0.361854    0.124453    0.13    0.2700   
Proanthocyanins       178.0    1.590899    0.572359    0.41    1.2500   
Color_intensity       178.0    5.058090    2.318286    1.28    3.2200   
Hue                   178.0    0.957449    0.228572    0.48    0.7825   
OD280                 178.0    2.611685    0.709990    1.27    1.9375   
Proline               178.0  746.893258  314.907474  278.00  500.5000   

                          50%       75%      max  
Cultivator              2.000    3.0000     3.00  
Alchol                 13.050   13.6775    14.83  
Malic_Acid              1.865    3.0825     5.80  
Ash                     2.360    2.5575     3.23  
Alcalinity_of_Ash      19.500   21.5000    30.00  
Magnesium              98.000  107.0000   162.00  
Total_phenols           2.355    2.8000     3.88  
Falvanoids              2.135    2.8750     5.08  
Nonflavanoid_phenols    0.340    0.4375     0.66  
Proanthocyanins         1.555    1.9500     3.58  
Color_intensity         4.690    6.2000    13.00  
Hue                     0.965    1.1200     1.71  
OD280                   2.780    3.1700     4.00  
Proline               673.500  985.0000  1680.00

In [7]:
wine.shape

(178, 14)

In [8]:
X = wine.drop('Cultivator',axis=1)
y = wine['Cultivator']

__Splitting the dataset into train test and validation dataset with test dataset of size 25% of main dataset.__

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

# Data Analysis

In [10]:
X_train.head()

Alchol  Malic_Acid   Ash  Alcalinity_of_Ash  Magnesium  Total_phenols  \
111   12.52        2.43  2.17               21.0         88           2.55   
119   12.00        3.43  2.00               19.0         87           2.00   
170   12.20        3.03  2.32               19.0         96           1.25   
142   13.52        3.17  2.72               23.5         97           1.55   
29    14.02        1.68  2.21               16.0         96           2.65   

     Falvanoids  Nonflavanoid_phenols  Proanthocyanins  Color_intensity   Hue  \
111        2.27                  0.26             1.22             2.00  0.90   
119        1.64                  0.37             1.87             1.28  0.93   
170        0.49                  0.40             0.73             5.50  0.66   
142        0.52                  0.50             0.55             4.35  0.89   
29         2.33                  0.26             1.98             4.70  1.04   

     OD280  Proline  
111   2.78      325  
119   3.05      564  
170   1.83      510  
142   2.06      520  
29    3.59     1035

We see there are 13 unique classes to classify wine.

# Standardize Data

In [11]:
scaler = StandardScaler()

In [12]:
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [13]:
X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

Standardization is a useful technique to transform attributes with a Gaussian distribution and differing means and standard deviations to a standard Gaussian distribution with a mean of 0 and a standard deviation of 1.The values for each attribute now have a mean value of 0 and a standard deviation of 1.

# Training the Multi-Layer Perceptron model

Training a 3 layer MLP with 'identity' as the activation function

In [19]:
mlp = MLPClassifier(hidden_layer_sizes=(13,12,11),max_iter=500,activation='identity')

In [20]:
mlp.fit(X_train,y_train)

MLPClassifier(activation='identity', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(13, 12, 11), learning_rate='constant',
       learning_rate_init=0.001, max_iter=500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [21]:
predictions = mlp.predict(X_test)

In [22]:
print(confusion_matrix(y_test,predictions))

[[ 8  0  0]
 [ 1 13  1]
 [ 0  0 13]]


In [23]:
print(classification_report(y_test,predictions))
print(accuracy_score(y_test,predictions))

             precision    recall  f1-score   support

          1       0.89      1.00      0.94         8
          2       1.00      0.87      0.93        15
          3       0.93      1.00      0.96        13

avg / total       0.95      0.94      0.94        36

0.9444444444444444


We see that after 500 iterations, accuracy score for the classification by this model is 94%.

We will now try to change the activation function and try to tune the number of hidden layers so that we will get a higher accuracy score along with higher appropriate classifications.We use 'tanh' activation function.

In [31]:
mlp1 = MLPClassifier(hidden_layer_sizes=(10,8,7),max_iter=2500,activation='identity')

In [32]:
mlp1.fit(X_train,y_train)

MLPClassifier(activation='identity', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(10, 8, 7), learning_rate='constant',
       learning_rate_init=0.001, max_iter=2500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [33]:
predictions = mlp1.predict(X_test)
print(confusion_matrix(y_test,predictions))
print(accuracy_score(y_test,predictions))

[[ 8  0  0]
 [ 0 15  0]
 [ 0  1 12]]
0.9722222222222222


Here, we observe that we are able to classify all the 13 classes of dataset with an accuracy of 97% approx which is more than the previous case after changing the number of neurons with activation function "identity"

In [37]:
mlp2 = MLPClassifier(hidden_layer_sizes=(6,8,7),max_iter=3500,activation='tanh')

In [38]:
mlp2.fit(X_train,y_train)

MLPClassifier(activation='tanh', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(6, 8, 7), learning_rate='constant',
       learning_rate_init=0.001, max_iter=3500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [39]:
predictions = mlp2.predict(X_test)
print(confusion_matrix(y_test,predictions))
print(accuracy_score(y_test,predictions))

[[ 8  0  0]
 [ 0 15  0]
 [ 0  1 12]]
0.9722222222222222


We observe that the accuracy of classification goes down to 97.22% approx when we use the tanh activation function which has a very slight difference than that of identity function.

In [65]:
mlp3 = MLPClassifier(hidden_layer_sizes=(1,1,1,1,1),max_iter=3500,activation='identity')

In [66]:
mlp3.fit(X_train,y_train)

MLPClassifier(activation='identity', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(1, 1, 1, 1, 1), learning_rate='constant',
       learning_rate_init=0.001, max_iter=3500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [67]:
predictions = mlp3.predict(X_test)
print(confusion_matrix(y_test,predictions))
print(accuracy_score(y_test,predictions))

[[ 8  0  0]
 [ 2 10  3]
 [ 0  1 12]]
0.8333333333333334


We observe that when we have 5 hidden layers with neuron values equal to 1, the classification accuracy is 83%. 

In [69]:
mlp4 = MLPClassifier(hidden_layer_sizes=(1,1,1,1,1),max_iter=3500,activation='tanh')

In [70]:
mlp4.fit(X_train,y_train)

MLPClassifier(activation='tanh', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(1, 1, 1, 1, 1), learning_rate='constant',
       learning_rate_init=0.001, max_iter=3500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [71]:
predictions = mlp4.predict(X_test)
print(confusion_matrix(y_test,predictions))
print(accuracy_score(y_test,predictions))

[[ 8  0  0]
 [ 1 14  0]
 [12  1  0]]
0.6111111111111112


We see that when we changed just the activation function considering everything else same as mlp3, we saw a dramatic decrease in the accuracy score.

# Conclusion

Overall, when we increase the number of hidden layers and change the values of neurons, the chances of missclassification increases as per in my dataset. Also considering the results of mlp3 and mlp4, we see that the type of activation function also matters a lot as the accuracy score went down fro 83% to 61%.